### PR TITLE
psc-bundle: Add source map output

### PR DESCRIPTION
**Pull request by nwolverson**

### CODE REVIEW

**Comment by nwolverson** [comment]
#2096

File: src/Language/PureScript/Bundle.hs [hunk 1]
```diff
@@ -612,26 +672,39 @@ codeGen optionsMainModule optionsNamespace ms =  renderToString (JSAstProgram (p
...
+-- | The bundling function.
+-- This function performs dead code elimination, filters empty modules
+-- and generates and prints the final Javascript bundle.
+bundle :: (MonadError ErrorMessage m)
+       => [(ModuleIdentifier, String)] -- ^ The input modules.  Each module should be javascript rendered from 'Language.PureScript.Make' or @psc@.
+       -> [ModuleIdentifier] -- ^ Entry points.  These module identifiers are used as the roots for dead-code elimination
+       -> Maybe String -- ^ An optional main module.
+       -> String -- ^ The namespace (e.g. PS).
+       -> m String
+bundle inputStrs entryPoints mainModule namespace = snd <$> bundleSM (map (\(a,b) -> (a,"",b)) inputStrs) entryPoints mainModule namespace Nothing
```
Comment (nwolverson) [inline src/Language/PureScript/Bundle.hs]: Not used in the no-source-maps code path, I guess this should be `Maybe`, or some better structuring generally

File: src/Language/PureScript/Bundle.hs [hunk 2]
```diff
@@ -98,10 +102,11 @@ data ModuleElement
   | Member JSStatement Bool String JSExpression [Key]
   | ExportsList [(ExportType, String, JSExpression, [Key])]
   | Other JSStatement
+  | Skip JSStatement
   deriving (Show)
 
 -- | A module is just a list of elements of the types listed above.
-data Module = Module ModuleIdentifier [ModuleElement] deriving (Show)
+data Module = Module ModuleIdentifier String [ModuleElement] deriving (Show)
```
Comment (paf31) [inline src/Language/PureScript/Bundle.hs]: Can this please use `FilePath`?

File: src/Language/PureScript/Bundle.hs [hunk 3]
```diff
@@ -490,26 +503,73 @@ isModuleEmpty (Module _ els) = all isElementEmpty els
...
+        zipWith3 (\file (pos :: Int) positions ->
+          map (\(porig, pgen) -> Mapping {
+                mapOriginal = Just (Pos (fromIntegral $ porig + 1) 0)
+              , mapSourceFile = Just (pathToFile file)
+              , mapGenerated = (Pos (fromIntegral $ pos + pgen) 0)
+              , mapName = Nothing
+              })
+            (offsets (0,0) (Right 1 : positions)))
+          moduleFns
+          (scanl (+) (3 + moduleLength [prelude]) (map (3+) moduleLengths))
```
Comment (paf31) [inline src/Language/PureScript/Bundle.hs]: Could you please add a short comment explaining the `3` here?

File: src/Language/PureScript/Bundle.hs [hunk 4]
```diff
@@ -612,26 +672,39 @@ codeGen optionsMainModule optionsNamespace ms =  renderToString (JSAstProgram (p
 -- | The bundling function.
 -- This function performs dead code elimination, filters empty modules
 -- and generates and prints the final Javascript bundle.
-bundle :: (MonadError ErrorMessage m)
-       => [(ModuleIdentifier, String)] -- ^ The input modules.  Each module should be javascript rendered from 'Language.PureScript.Make' or @psc@.
+bundleSM :: (MonadError ErrorMessage m)
+       => [(ModuleIdentifier, String, String)] -- ^ The input modules.  Each module should be javascript rendered from 'Language.PureScript.Make' or @psc@.
```
Comment (paf31) [inline src/Language/PureScript/Bundle.hs]: Same here (`FilePath`)?

File: src/Language/PureScript/Bundle.hs (hunk 1 above)
Comment (paf31) [inline src/Language/PureScript/Bundle.hs]: Maybe use `internalError` if it's not used.

File: src/Language/PureScript/Bundle.hs (hunk 3 above)
Comment (nwolverson) [inline src/Language/PureScript/Bundle.hs]: Yes, if I can reconstruct why - wish I did so in the first place!

**Comment by nwolverson** [comment]
Thanks for the review @paf31, will get to this post-xmas

File: psc-bundle/Main.hs [hunk 5]
```diff
@@ -30,32 +35,42 @@ import qualified Options.Applicative as Opts
...
     mid <- guessModuleIdentifier filename
-    length js `seq` return (mid, js)                                            -- evaluate readFile till EOF before returning, not to exhaust file handles
+    length js `seq` return (mid, filename, js)                                            -- evaluate readFile till EOF before returning, not to exhaust file handles
 
   let entryIds = map (`ModuleIdentifier` Regular) optionsEntryPoints
 
-  bundle input entryIds optionsMainModule optionsNamespace
+  currentDir <- liftIO getCurrentDirectory
+  let outFile = if optionsSourceMaps then fmap (currentDir </>) optionsOutputFile else Nothing
+  bundleSM input entryIds optionsMainModule optionsNamespace outFile
```
Comment (paf31) [inline psc-bundle/Main.hs]: Travis says the issue is here, since `input` needs `Maybe FileName` in its second argument.

Comment (nwolverson) [inline psc-bundle/Main.hs]: Guess I don't know what I built after making changes but it wasn't psc-bundle! Fixed  & squashed.

**Comment by paf31** [comment]
Thanks!

**Pull request merged**